Additions for RotatedBricks 2

mar-celine opened the pull request:
## Proposed changes

Additions for RotatedBricks 1 gave the user the ability to produce the proper global corner numbering
for any n x m x l sized edifice of Blocks.

Additions for RotatedBricks 2 gives the user the ability to produce the corresponding CoordinateMaps for these Blocks.

Allows for the rotation of blocks "in place" within the context of domain creation:
Given a (n,m,l) edifice of aligned blocks, a user can select one or more blocks and change their orientation relative to the edifice - the corners algorithm will then deduce the correct relative orientations between all of these blocks and their neighbors.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

kidder commented on src/Domain/OrientationMap.hpp:
```diff
@@ -67,10 +67,20 @@ class OrientationMap {
   /// The corresponding Orientation of the host in the frame of the neighbor.
   OrientationMap<VolumeDim> inverse_map() const noexcept;
 
+  /// The correct ordered graph of global corners for a block having
```
free your function (into DomainHelpers)

kidder commented on src/Domain/OrientationMap.hpp:
```diff
@@ -67,10 +67,20 @@ class OrientationMap {
   void pup(PUP::er& p) noexcept;  // NOLINT
 
  private:
+  /// The corresponding logical corners in the neighbor.
```
remove

kidder reviewed (changes requested):
split first commit into move then edits

mar-celine committed:
Additions to DomainHelpers:

- Extend VolumeCornerIterator to hold Directions

- Extend discrete_rotation to act on corner numberings

mar-celine committed:
Add maps_for_rectilinear_domains

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -147,9 +147,9 @@ std::vector<std::array<size_t, 8>> corners_for_biradially_layered_domains(
-  explicit VolumeCornerIterator(size_t initial_local_corner_number) noexcept
-      : local_corner_number_(initial_local_corner_number) {}
+  VolumeCornerIterator() noexcept {setup_from_local_corner_number();}
+  explicit VolumeCornerIterator(size_t initial_local_corner_number) noexcept 
```
whitespace

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -194,12 +188,27 @@ class VolumeCornerIterator {
   const std::array<double, VolumeDim>& coords_of_corner() const noexcept {
     return coords_of_corner_;
   }
+  const std::array<Direction<VolumeDim>, VolumeDim>& directions_of_corner() const noexcept{
```
... and put it here

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -194,12 +188,27 @@ class VolumeCornerIterator {
+   for (size_t i = 0; i < VolumeDim; i++) {
+      gsl::at(coords_of_corner_, i) =
+          2.0 * get_nth_bit(local_corner_number_, i) - 1.0;
+      const auto side = 
```
inline this on the next line and use the array on 202

kidder commented on tests/Unit/Domain/Test_DomainHelpers.cpp:
```diff
@@ -586,9 +586,13 @@ void test_vci_1d() {
   CHECK(vci);
   CHECK(vci() == std::array<Side, 1>{{Side::Lower}});
   CHECK(vci.coords_of_corner() == std::array<double, 1>{{-1.0}});
+  CHECK(vci.directions_of_corner() ==
```
add test that catches your previous bug

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -142,14 +144,28 @@ std::vector<std::array<size_t, 8>> corners_for_biradially_layered_domains(
+///
+/// Returns the correct ordering of global corner numbers for a block
+/// having this orientation relative to an aligned edifice of blocks,
+/// given the ordered graph the block would have if it were aligned.
```
ordered graph -> corner numbering

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -142,14 +144,28 @@ std::vector<std::array<size_t, 8>> corners_for_biradially_layered_domains(
+/// \ingroup ComputationalDomainGroup
+/// \brief Permutes the corner numbers of an n-cube.
+///
+/// Returns the correct ordering of global corner numbers for a block
```
give a concrete example, and what it is useful for

kidder commented on src/Domain/DomainHelpers.hpp:
```diff
@@ -188,16 +204,17 @@ class VolumeCornerIterator {
-      const auto side = 
-          2 * get_nth_bit(local_corner_number_, i) - 1 == 1 ? Side::Upper
-                                                            : Side::Lower;
+      const auto side = 2 * get_nth_bit(local_corner_number_, i) - 1 == 1
```
merge formatting changes into previous commit

kidder commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -664,6 +664,46 @@ corners_for_rectilinear_domains(
+  // compute the mapped logical corners, as
+  // they are the indices into the global corners.
+  const std::array<size_t, two_to_the(VolumeDim)> mapped_logical_corners =
+      [&]() {
```
capture what you use

kidder replied:
noexcept

kidder commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -664,6 +664,46 @@ corners_for_rectilinear_domains(
+          for (size_t i = 0; i < VolumeDim; i++) {
+            const auto new_direction =
+                orientation(gsl::at(directions_of_logical_corner, i));
+            gsl::at(directions_of_mapped_corner, i) = new_direction;
```
inline new_direction

kidder commented on tests/Unit/Domain/Test_DomainHelpers.cpp:
```diff
@@ -818,3 +818,49 @@ SPECTRE_TEST_CASE("Unit.Domain.DomainHelpers.CornersForRectilinearDomains",
+            OrientationMap<2>(std::array<Direction<2>, 2>{
+                {Direction<2>::lower_eta(), Direction<2>::lower_xi()}}),
+            std::array<size_t, 4>{{0, 1, 2, 3}}));
+  CHECK(std::array<size_t, 8>{{9, 0, 12, 3, 10, 1, 13, 4}} ==
```
blank line here

kidder commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -664,6 +666,77 @@ corners_for_rectilinear_domains(
+std::unique_ptr<CoordinateMapBase<Frame::Logical, TargetFrame, 3>>
+product_of_1d_maps(std::array<Map, 3>& maps) {
+  return make_coordinate_map_base<Frame::Logical, TargetFrame>(
+      CoordinateMaps::ProductOf3Maps<Map, Map, Map>(maps[0], maps[1], maps[2]));
```
use tuples instead of arrays

kidder reviewed (commented):
squash
